File: src/main/java/com/github/sorusclient/client/ui/framework/Scroll.java
package com.github.sorusclient.client.ui.framework;

public class Scroll extends List {

    public Scroll(int type) {
        super(type);
        this.runtime = new Runtime();

        this.addStoredState("scroll");
        this.setOnInit(state -> state.getSecond().put("scroll", 0.0));
        this.setOnScroll(state -> {
            state.getSecond().put("scroll", (double) state.getSecond().get("scroll") + state.getFirst() * 3);
        });
    }

    public class Runtime extends List.Runtime {

        private double prevYLocation;

        @Override
        public void render(double x, double y, double width, double height) {
            super.render(x, y, width, height);
            this.prevYLocation = this.yLocation + height / 2;
        }

        @Override
        protected void renderChild(Component.Runtime childRuntime, double x, double y, double width, double height) {
            double scroll = (double) this.getState("scroll");
            scroll = Math.max(scroll, -(this.prevYLocation - this.height + this.getChildren().get(0).getRuntime().getPadding()));
            scroll = Math.min(scroll, 0);
            this.setState("scroll", scroll);
            super.renderChild(childRuntime, x, y + scroll, width, height);
        }

    }

}

File: src/main/kotlin/com/github/sorusclient/client/ui/framework/Scroll.kt
/*
 * Copyright (c) 2022. Danterus
 * Copyright (c) 2022. Sorus Contributors
 *
 * SPDX-License-Identifier: MPL-2.0
 */

package com.github.sorusclient.client.ui.framework

import kotlin.math.max
import kotlin.math.min

class Scroll(type: Int) : List(type) {

    init {
        runtime = Runtime()
        storedState += "scroll"
        onScroll = { state ->
            state.second["scroll"] = state.second.getOrDefault("scroll", 0.0) as Double + state.first * 3
        }
        scissor = true
    }

    inner class Runtime : List.Runtime() {
        private var prevYLocation = 0.0

        override fun render(x: Double, y: Double, width: Double, height: Double) {
            super.render(x, y, width, height)
            prevYLocation = yLocation + height / 2
        }

        override fun renderChild(
            childRuntime: Component.Runtime?,
            x: Double,
            y: Double,
            width: Double,
            height: Double
        ) {
            var scroll = getState("scroll") as Double?
            if (scroll == null || scroll.isNaN()) {
                scroll = 0.0
            }
            scroll = max(scroll, -(prevYLocation - this.height + children[0].runtime.topPadding))
            scroll = min(scroll, 0.0)
            setState("scroll", scroll)
            super.renderChild(childRuntime, x, y + scroll, width, height)
        }
    }
}